### ISSUE TITLE
Name does not update in homebridge

### ISSUE THREAD

**Issue by smulle48** [itavero/homebridge-z2m#76]
Hi

I don't know if the problem is in this plugin.

The problem is that if I have homebridge running while adding a new item in z2m, the friendly name does not update in homebridge and in the Home app.

As you can see in the screenshots, I have added a bulb in my daughters room. After inclusion, I have changed the name in z2m, but as homebridge was running while including the bulb, the IEEE address was used as the name. As you can see in homebridge, the name is indeed changed, but not used. Even if I manually change the name in homebridge, it does not change in the Home app and the only way to make it change (except to manually change it) is to reset homebridge or make sure homebridge is not running while including devices.

![Skærmbillede 2021-02-23 kl  13 00 34](https://user-images.githubusercontent.com/24297824/108841636-5d156100-75d8-11eb-9e16-d78ebe16f638.png)
![Skærmbillede 2021-02-23 kl  13 00 16](https://user-images.githubusercontent.com/24297824/108841643-61417e80-75d8-11eb-94fb-40f61a84f205.png)
![Skærmbillede 2021-02-23 kl  13 10 07](https://user-images.githubusercontent.com/24297824/108841713-7d452000-75d8-11eb-8007-b45f9e2b3f5c.png)

**Comment by itavero** [comment]
Updating the display name in HomeKit automatically doesn't appear to be possible (gave it a few attempts, but without any luck). Personally I also don't want/need this, as my friendly names in Zigbee2MQTT also have some stuff in there to identify the type of device and location, but I don't need that in the name for HomeKit.

You can ofcourse still rename it in the Home app itself.

**Comment by smulle48** [comment]
> Updating the display name in HomeKit automatically doesn't appear to be possible (gave it a few attempts [in the code](https://github.com/itavero/homebridge-z2m/blob/7317baed07bded605b19c2c9772ee9f3531f09af/src/platformAccessory.ts#L237), but without any luck). Personally I also don't want/need this, as my friendly names in Zigbee2MQTT also have some stuff in there to identify the type of device and location, but I don't need that in the name for HomeKit.
> 
> 
> 
> You can ofcourse still rename it in the Home app itself.

When I used deCONZ in Jeedom with their homebridge plugin, it updated the name straight away

**Comment by itavero** [comment]
Can you maybe add a link to that plugin here?
I'll see if I can peek at the code and add an option to enable this functionality (as mentioned, personally I do not want this to happen).

**Comment by smulle48** [comment]
> Can you maybe add a link to that plugin here?
> 
> I'll see if I can peek at the code and add an option to enable this functionality (as mentioned, personally I do not want this to happen).

The problem with Jeedom is that most of the plugins are paid plugins so they are on some private repo or something, but I think this is the plug-in :-)

https://github.com/NebzHB/homebridge-jeedom

Maybe it could be as option (turned off by default)

Just added all the devices in the house, just to realise that I had no idea which device was in which room :-)

It's a real pain in the a** to rename 100+ devices :-)

**Comment by itavero** [comment]
As you can tell in your screenshot of Homebridge Config UI X, the name for the accessory is actually what you expect (a bit further down in the table).
However, the name for the service is not. I think this is what needs to be done, to get the behavior you are requesting.
The Jeedom plugin is partially with French comments, but a quick glance at it, suggests that this is also what they are doing.

Currently the service uses the friendly name up on creation (plus some additional info in some specific cases).
However, the name of the service should probably not be set explicitly by the plugin, unless there are multiple endpoints or for instance when different `ServiceHandler`s can implement the same service (i.e. water leak vs gas leak). I think/hope that HomeKit will automagically use the name of the accessory the service belongs to in that case, but I'm not 100% sure.

@smulle48  Just wondering, if you restart Homebridge, do you still see the incorrect names?
I might be able to make it in such a way that restarting Homebridge at least updates the names of the services. Not the best solution maybe, but it does look like it is the easiest to implement.

**Comment by smulle48** [comment]
> As you can tell in your screenshot of Homebridge Config UI X, the name for the accessory is actually what you expect (a bit further down in the table).
> 
> However, the name for the service is not. I think this is what needs to be done, to get the behavior you are requesting.
> 
> The Jeedom plugin is partially with French comments, but a quick glance at it, suggests that this is also what they are doing.
> 
> 
> 
> Currently the service uses the friendly name up on creation (plus some additional info in some specific cases).
> 
> However, the name of the service should probably not be set explicitly by the plugin, unless there are multiple endpoints or for instance when different `ServiceHandler`s can implement the same service (i.e. water leak vs gas leak). I think/hope that HomeKit will automagically use the name of the accessory the service belongs to in that case, but I'm not 100% sure.
> 
> 
> 
> @smulle48  Just wondering, if you restart Homebridge, do you still see the incorrect names?
> 
> I might be able to make it in such a way that restarting Homebridge at least updates the names of the services. Not the best solution maybe, but it does look like it is the easiest to implement.

The screenshots are with homebridge running, after I have renamed it in z2m.

In the home app, it's also using the ieee as name :-(

Yes, my first thought was that I could just restart homebridge, but that did not change anything.

But a cache reset (they don't recommend it) fixes the names.

Now that I know the "limitations" I just shut down homebridge while including, but sometimes I forget it :-)

If it's just added as an extra feature you have to enable manually, is totally fine with me :-) (ie. enable to update names at reboot)

**Comment by smulle48** [comment]
No worries 😉 

Now that I know it's not due to a bug in you nice plugin.

Yes, after changing the name in z2m, name is changed, but the name in the top is what's used in the home app

**Comment by itavero** [comment]
Another developer on the Homebridge Discord mentioned that someone reported a bug that their plugin always overwrites the name in Homebridge.
It's more or less the opposite of this, but it did gave me some insight on how to implement this.

Currently for the Accessory I actively set the name and for the services I only set it on creation.
Instead I should at a "get" callback for the Name characteristic so it can always return an up to date default name when HomeKit requests it, similar to what is done in [this plugin](https://github.com/OpenWonderLabs/homebridge-switchbot-openapi/blob/18213f1c639920540baf7f694c0f86cd1b9cb712/src/irdevices/tvs.ts#L64).

I'll try to have a go it implementing this when I have time.

**Comment by itavero** [comment]
Ignore the previous comment. I don't have to look at the Configured Name.

I think I've fixed it now by explicitly updating the _Name_ characteristic of all services (based on the `friendly_name`).
I've only tested it on my development environment, but it does seem to work (and to not overwrite the name if it has been reconfigured from the Home app.

The Homebridge Config UI X overview seems to update immediately, but for the Home app to pick up the change it seems that I have to restart Homebridge.
I think that's a minor issue and would still help a lot for users in the same situation as @smulle48.

I'm going to do a bit more testing ;)
